src/main/java/org/opensearch/knn/training/VectorReader.java

```diff
@@ -0,0 +1,162 @@
+                .execute(vectorReaderListener);
+    }
+
+    @SuppressWarnings("unchecked")
+    private ActionListener<SearchResponse> getVectorReaderListener(String fieldName,
+                                                                   final int maxVectorCount,
+                                                                   final int collectedVectorCount,
+                                                                   Consumer<List<Float[]>> vectorConsumer,
+                                                                   ActionListener<SearchResponse> listener) {
+        return ActionListener.wrap(searchResponse -> {
+            // Get the vectors from the search response
+            // Either add the entire set of returned hits, or maxVectorCount - collectedVectorCount hits
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            int vectorsToAdd = Integer.min(maxVectorCount - collectedVectorCount, hits.length);
+            List<Float[]> trainingData = new ArrayList<>();
+
+            for (int i = 0; i < vectorsToAdd; i++) {
+                trainingData.add(((List<Double>) hits[i].getSourceAsMap().get(fieldName)).stream()
+                        .map(Double::floatValue)
+                        .toArray(Float[]::new));
+            }
+
+            final int updatedVectorCount = collectedVectorCount + trainingData.size();
+
+            // Do something with the vectors
+            vectorConsumer.accept(trainingData);
+
+            if (vectorsToAdd <= 0 || updatedVectorCount >= maxVectorCount) {
+                listener.onResponse(searchResponse);
+            } else {
+                ActionListener<SearchResponse> vectorReaderListener = getVectorReaderListener(fieldName, maxVectorCount,
+                        updatedVectorCount, vectorConsumer, listener);
+
+                // Create a new search that starts where the last search left off
+                createSearchScrollRequestBuilder(searchResponse).execute(vectorReaderListener);
+            }
+        }, listener::onFailure);
+    }
+
+    private SearchRequestBuilder createSearchRequestBuilder(String indexName, String fieldName, int resultSize) {
+        ExistsQueryBuilder queryBuilder = new ExistsQueryBuilder(fieldName);
+
+        SearchRequestBuilder searchRequestBuilder = client.prepareSearch(indexName);
+        searchRequestBuilder.setScroll(scrollTime);
+        searchRequestBuilder.setQuery(queryBuilder);
+        searchRequestBuilder.setSize(resultSize);
+
+        // We are only interested in reading vectors from a particular field
+        searchRequestBuilder.setFetchSource(fieldName, null);
+
+        return searchRequestBuilder;
+    }
+
+    private SearchScrollRequestBuilder createSearchScrollRequestBuilder(SearchResponse searchResponse) {
+        SearchScrollRequestBuilder searchScrollRequestBuilder = client.prepareSearchScroll(searchResponse.getScrollId());
```

Comment:
https://www.elastic.co/guide/en/elasticsearch/reference/6.8/search-request-scroll.html recommends we use

```
GET /_search?scroll=1m
{
  "sort": [
    "_doc"
  ]
}
```

Do you want to add that option if the sort order does not matter to you?

Reply:
Yes, I good catch. I will update.